### PR TITLE
SWATCH-1924: Support querying prometheus/rhelemeter when "product" label has engIds

### PR DESCRIPTION
**Pull request by lindseyburnett**
Jira issue: SWATCH-1924

**Context**
Getting creative with how we can detect these RHEL payg products with rhelemeter, since the role available on the hosts isn't unique enough to differentiate between different variants.

This change enables hosts to report a list of comma-separated engineering Ids in its "product" label. We'll detect comma-separated engineering Ids and set the swatch event.productIds field accordingly.

During hourly tally, we already match events to swatch subscription definitions by using event.role or event.productIds anyway.

# summary

## promql query changes
Introduced a new promql query template.  There's a couple of differences than the default template
* removes "metadataMetric" parameter
* adds "productLabelRegex" parameter.
* removes instanceKey parameter (note that instanceKey still needs to be defined on the product config files because it's being used programmatically elsewhere)

The rhel-for-x86-els-payg product configuration yaml has the regex defined as any comma-separated list that includes "204".

## Event message updates
Leverage changes made in #2844 and populate `productTag` field when emitting a metric Event on the service instance topic.  This is a little redundant, because the ingestion of metric Events (takes place in swatch-tally) looks up the productTag based on productIds & role and will re-set the value.  There's room for improvement here to potentially stop looking up productTag if an Event has a productTag set in the payload already. 

Populate `productIds` field when emitting a metric Event on the service instance topic.  The value comes from the prometheus response in the "product" payload.  As previously mentioned, during Event ingestion there's a productTag lookup.  Since there isn't a role that corresponds to rhel-for-x86-els-payg, that productTag can only be resolved by having productIds present.

## misc

While working on these changes, I encountered an issue with the timeframes that recording rules are evaluated.  There are non-trivial updates required for this, as detailed in SWATCH-2075.  As a workaround, I've set the `metric` field in the product config file to use the expression that backs the recording rule.  This workaround is to enable testing, and isn't a long-term solution.  We ultimately don't want to be attempting to maintain usage queries that are defined/crafted/owned by other product teams.

# expected behavior
* a POST request to `/api/swatch-metrics/v1/internal/metering/rhel-for-x86-els-payg` successfully retrieves data from rhelemeter
* swatch-metrics (or swatch-metrics-rhel if you're in EE) emits Event messages for the metrics returned from rhelemeter
* swatch-tally successfully consumes those message, resulting in entries in the rhsm-subscriptions.events database table
* a PUT request to `/api/rhsm-subscriptions/v1/internal/rpc/tally/snapshots` should result in snapshots

# test steps

these are written for local, but similar approach should work in ephemeral by port forwarding pods.  token refresher will already be attached to the swatch-metrics-rhel pod.

### 1. start up a token-refresher for stage rhelemeter

Secrets can be found in vault under _secrets/insights-stage/rhsm-stage/rhel-token-refresher_

```
podman run --name=token-refresher-$1 --rm -ti -p 8082:8080 \
  quay.io/observatorium/token-refresher:master-2023-09-20-f5e3403 \
  --scope="$SCOPE" \
  --oidc.client-secret="$CLIENT_SECRET" \
  --oidc.client-id="$CLIENT_ID" \
  --oidc.issuer-url="$ISSUER_URL" \
  --url="https://observatorium-mst.api.stage.openshift.com/api/metrics/v1/rhel"
  ```

### 2. start up swatch-metrics-rhel
```bash
EVENT_SOURCE=rhelemeter PROM_URL="http://localhost:8082/api/v1/" ./gradlew :swatch-metrics:quarkusDev
```
### 3. kick off metering for test org
```bash
curl -X POST --location "http://localhost:8000/api/swatch-metrics/v1/internal/metering/rhel-for-x86-els-payg?orgId=16787820" \
    -d 'accept: */*
x-rh-swatch-synchronous-request: false'
```

if you're running this at DEBUG log level, you should see some log statements containing details about the range query being done against prometheus - promql, step, and start/end ranges.

### 5. start up swatch-tally and verify it consumes the messages
`SERVER_PORT=8001 DEV_MODE=true ./gradlew :bootRun`

you should see messages getting consumed after it starts up.

after the events are consumed, check the `data` column in the `events` table that the information from the Event message persisted properly (had issues at one point with productTag getting blown away before persisting)

### 6. kick off hourly tally
will have to adjust based on timestamps you're doing this

```
curl -X 'PUT' \
'http://localhost:8001/api/rhsm-subscriptions/v1/internal/rpc/tally/snapshots?begin=2024-01-08T19%3A00%3A00Z&end=2024-01-08T23%3A00%3A00Z' \
-H 'accept: application/json' \
-H 'x-rh-swatch-psk: placeholder'
```

### CODE REVIEW

File: swatch-metrics/src/main/java/com/redhat/swatch/metrics/service/PrometheusMeteringController.java [hunk 1]
```diff
@@ -198,11 +202,19 @@ private void sendEventFromData(
     String product = labels.get("product");
     String resourceName = labels.get("resource_name");
 
+    // TODO update this note
```
Comment (lindseyburnett) [inline swatch-metrics/src/main/java/com/redhat/swatch/metrics/service/PrometheusMeteringController.java:205]: Looking for some assistance on how to wordsmith this note update.

File: swatch-product-configuration/src/main/resources/subscription_configs/RHEL/RHEL_for_x86_els_payg.yaml [hunk 2]
```diff
@@ -29,9 +27,8 @@ metrics:
     awsDimension: vCPU_hours
     azureDimension: vcpu_hours
     prometheus:
-      queryKey: default
+      queryKey: rhelemeter
       queryParams:
+        productLabelRegex: .*(^|,)(204)($|,).*
```
Comment (Sgitario) [inline swatch-product-configuration/src/main/resources/subscription_configs/RHEL/RHEL_for_x86_els_payg.yaml:32]: Can you add a comment about what the intention with this value is and where the value 204 comes from?
This is used by `product=~"#{metric.prometheus.queryParams[productLabelRegex]}"`, so I guess you want to only fetch metrics for product that contains the product ID 204, right?

**Review by Sgitario — COMMENTED** [review]
Changes look good to me. 
The test steps to verify the changes worked for me.
I only added a few minor comments.

**Review by ssidhaye — APPROVED** [review]
- a metering POST request successfully retrieves data from rhelemeter
- swatch-metrics (or swatch-metrics-rhel if you're in EE) emits Event messages for the metrics returned from rhelemeter
- swatch-tally comsumes messages from above steps and populates events table (product_tag and product_ids) 
- a PUT request to /api/rhsm-subscriptions/v1/internal/rpc/tally/snapshots results in snapshots

**Pull request merged**